src/comps/qb.cc

```diff
@@ -0,0 +1,407 @@
+
+            if (orth_i_err > 1.0e-10)
+            {
+                // Lost orthonormality of Q_i
+                RandLAPACK::comps::util::qb_resize( m, n, Q, B, k, curr_sz);
+                QB::termination = 4;
+                return;
+            }
+        }
+
+        // No need to reorthogonalize on the 1st pass
+        if(curr_sz != 0)
+        {
+            // Q_i = orth(Q_i - Q(Q'Q_i))
+            gemm(Layout::ColMajor, Op::Trans, Op::NoTrans, next_sz, block_sz, m, 1.0, Q_dat, m, Q_i_dat, m, 0.0, QtQi_dat, k);
+            gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, block_sz, next_sz, -1.0, Q_dat, m, QtQi_dat, k, 1.0, Q_i_dat, m);
+
+            // If CholQR succeeded in the rangefinder
+            if (!QB::RF_Obj.Orth_Obj.chol_fail)
+            {
+                // No need to perform failure check here, as RF1 will notify in advance
+                // Done via CholQR
+                QB::Orth_Obj.call(m, block_sz, Q_i);
+                // Performing the alg twice for better orthogonality	
+                QB::Orth_Obj.call(m, block_sz, Q_i);
+            }
+            else
+            {
+                QB::Orth_Obj.decision_orth = 1;
+                // Done via regular LAPACK's QR
+                QB::Orth_Obj.call(m, block_sz, Q_i);
+            }
+        }
+        //B_i = Q_i' * A
+        gemm<T>(Layout::ColMajor, Op::Trans, Op::NoTrans, block_sz, n, m, 1.0, Q_i_dat, m, A_cpy_dat, m, 0.0, B_i_dat, block_sz);
+
+        // Updating B norm estimation
+        T norm_B_i = lange(Norm::Fro, block_sz, n, B_i_dat, block_sz);
+        norm_B = hypot(norm_B, norm_B_i);
+        // Updating approximation error
+        prev_err = approx_err;
+        approx_err = sqrt(abs(norm_A - norm_B) * (norm_A + norm_B)) / norm_A;
+
+        // Early termination - handling round-off error accumulation
+        if ((curr_sz > 0) && (approx_err > prev_err))
+        {
+            // Early termination - error growth
+            // Resizing the output arrays
+            RandLAPACK::comps::util::qb_resize( m, n, Q, B, k, curr_sz);
+            QB::termination = 2;
+            return;
+        } 
+
+        // Update the matrices Q and B
+        lacpy(MatrixType::General, m, block_sz, Q_i_dat, m, Q_dat + (m * curr_sz), m);	
+        lacpy(MatrixType::General, block_sz, n, B_i_dat, block_sz, B_dat + curr_sz, k);
+
+        //char name_3[] = "QB1  output Q";
+        //RandBLAS::util::print_colmaj(m, k, Q.data(), name_3);
+
+        //char name_4[] = "QB1  output B";
+        //RandBLAS::util::print_colmaj(k, n, B.data(), name_4);
+
+        if(QB::orth_check)
+        {
+            gemm(Layout::ColMajor, Op::Trans, Op::NoTrans,
+                k, k, m,
+                1.0, Q_dat, m, Q_dat, m,
+                0.0, Q_gram_dat, k
+            );
+            for (int oi = 0; oi < next_sz; ++oi) {
+                Q_gram_dat[oi * k + oi] -= 1.0;
+            }
+            T orth_err = lange(Norm::Fro, k, k, Q_gram_dat, k);
```

Comment:
It would be nice to only write to the first ``next_sz * next_sz`` entries of ``Q_gram_dat``. We can just lie about it's leading dimension.